### ISSUE TITLE
Add support for 'Cucumber expressions'

### ISSUE THREAD

**Issue by pfrenssen** [Behat/Behat#1254]
The reference implementations for Gherkin has adopted 'Cucumber expressions' which are a type based argument substitution in step definitions, similar to ours but using a different syntax. Ref. https://cucumber.io/docs/cucumber/cucumber-expressions/

We currently allow this transformation in step definitions (ref #863):

```
@Then I should be a user named :name
```

In order to increase compatibility with the reference Gherkin implementation, we can also support this transformation:

```
@Then I should be a user named {string}
```

**Comment by ciaranmcnulty** [comment]
Yes I'd like to do this long-term. Probably first step is a PHP implementation in cucumber/cucumber-expressions
